### PR TITLE
fs: Improve documentation for FS API

### PR DESCRIPTION
**Pull request by de-nordic**
Simple improvements in text and formatting of the documentation
the the FS API.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

### CODE REVIEW

**Review by nvlsianpu — APPROVED** [review]

File: include/fs/fs.h [hunk 1]
```diff
@@ -203,24 +204,28 @@ struct fs_file_system_t {
...
- * Opens or creates, if does not exist, file depending on flags provided
- * and associates a stream with it.
+ * Opens or possibly creates a file and associates a stream with it.
  *
- * @param zfp Pointer to file object
- * @param file_name The name of file to open
+ * @details
+ * @p flags can be 0 or a binary combination of one or more of the following
+ * identifiers:
+ *   - <tt>FS_O_READ</tt>  open for read
```
Comment (pabigot) [inline include/fs/fs.h:212]: FWIW `@c FS_O_READ` would be the standard doxygen way to style code text.  `<tt>FS_O_READ | FS_O_WRITE</tt>` would be necessary for a multi-token sequence.

File: include/fs/fs.h [hunk 2]
```diff
@@ -294,15 +300,15 @@ ssize_t fs_read(struct fs_file_t *zfp, void *ptr, size_t size);
...
  * @param size Number of bytes to be write
  *
- * @return Number of bytes written. On success, it will be equal to the number
- * of bytes requested to be written. Any other value, indicates an error. Will
- * return -ERRNO code on error.
- * In the case where -ERRNO is returned, the file pointer will not be
+ * @retval >=0 number of bytes written, on success;
+ * @retval <0 negative errno code on error.
+ *
+ * @return In the case where error is returned, the file pointer will not be
```
Comment (pabigot) [inline include/fs/fs.h:306]: I generally prefer to see descriptions like this and the one for `fs_read` up in the description of how the function behaves, and in the return value summary just have `count of bytes written` maybe with `as described above` if that doesn't seem enough.

File: include/fs/fs.h [hunk 3]
```diff
@@ -341,34 +348,34 @@ off_t fs_tell(struct fs_file_t *zfp);
  *
  * Truncates the file to the new length if it is shorter than the current
  * size of the file. Expands the file if the new length is greater than the
- * current size of the file. The expanded region would be filled with zeroes.
+ * current size of the file.  The expanded region would be filled with zeroes.
```
Comment (pabigot) [inline include/fs/fs.h:351]: "Two spaces after a period" is a style preferred by people who learned to type on typewriters and stick with fixed-width fonts on computers (raises hand).  While I use it and [science supports it](https://www.theatlantic.com/science/archive/2018/05/two-spaces-after-a-period/559304/) I don't think existing text with one space should be changed unless that becomes a project-wide standard.  Or at least a standard within the file, in which case it should probably be done in a separate white-space--only commit as is preferred for styling code.

**Review by pabigot — APPROVED** [review]
I'm okay with this as is, though putting detailed description of how the return value is determined in the return value summary rather than the documentation of what the function does is borderline.

The rest is just opinion/personal preference.  So fastest path is go run with it as-is.

**Comment by de-nordic** [comment]
> 
> 
> I'm okay with this as is, though putting detailed description of how the return value is determined in the return value summary rather than the documentation of what the function does is borderline.
> 
> The rest is just opinion/personal preference. So fastest path is go run with it as-is.

OK, I will move the description.

File: include/fs/fs.h (hunk 1 above)
Comment (de-nordic) [reply include/fs/fs.h:212]: Will fix.

File: include/fs/fs.h (hunk 3 above)
Comment (de-nordic) [reply include/fs/fs.h:351]: Actually I have altered the description and undo it not enough times and this is leftover.
But I can do unification of all the spaces after period into one style, at all this PR claims to improve the documentation...
@pabigot which way are we going then? "Two spaces after period" is it?

Comment (pabigot) [reply include/fs/fs.h:351]: Yeah, but we don't have consensus on whether two or one spaces is an improvement.  My opinion: if you didn't touch an adjacent sentence for some other reason, leave the spacing alone.  If you did, use the spacing that's used in the original version of the comment.  If it was inconsistent before, pick whichever was most often used for the content you changed, and leave the others alone (even if they remain inconsistent).

Minimal changes necessary to accomplish the task are generally preferred.

Comment (de-nordic) [reply include/fs/fs.h]: OK

File: include/fs/fs.h [hunk 4]
```diff
@@ -442,66 +447,66 @@ int fs_closedir(struct fs_dir_t *zdp);
...
- * @brief Mount point read entry
+ * @brief Get path of mount point at index
  *
- * Read mount point entry
+ * Function iterates through list of mount points and returns directory name
+ * of mount point at given index.
+ * On success @p index is incremented and @p name is set to mount directory
+ * name.  If mount point with given @p index does not exist, @p name will be
+ * assigned @c NULL value
  *
```
Comment (pabigot) [inline include/fs/fs.h:477]: "~F~*This f*unction iterates through *the* list of mount points and returns *the* directory name of *the* mount point at *the* given index. On success @<>p index is incremented and @<>p name is set to *the* mount directory name.  If *a* mount point with *the* given @<>p index does not exist, @<>p name will be ~assigned~ *set to* @<>c NULL~ value~."

Yes, English is verbose.

File: include/fs/fs.h [hunk 5]
```diff
@@ -442,66 +447,66 @@ int fs_closedir(struct fs_dir_t *zdp);
...
- * @retval 0 Success
- * @retval -ERRNO errno code if error
+ * @retval 0 on success;
+ * @retval <0 negative errno code on error.
  */
 int fs_stat(const char *path, struct fs_dirent *entry);
 
 /**
- * @brief Retrieves statistics of the file system volume
+ * @brief statistics of the file system volume
```
Comment (pabigot) [inline include/fs/fs.h:501]: I think the original is better as it says what the function *does*.

**Review by pabigot — CHANGES_REQUESTED** [review]
Some edits related to language.  I think this is overall an improvement, and is easier to verify without the whitespace changes.

File: include/fs/fs.h [hunk 6]
```diff
@@ -201,56 +203,59 @@ struct fs_file_system_t {
...
- *   FS_O_CREATE  create file if it does not exist
- *   FS_O_APPEND  move to end of file before each write
- *
- * @retval 0 Success
- * @retval -EINVAL when bad file name is given
- * @retval -NOENT when file path is not possible (bad mount point)
- * @retval other negative error code, depending on file system back-end.
+ * @retval 0 on success;
+ * @retval -EINVAL when bad file name is given;
+ * @retval -NOENT when file path is not possible (bad mount point);
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:228]: Eh, will fix.

File: include/fs/fs.h [hunk 7]
```diff
@@ -260,115 +265,117 @@ int fs_unlink(const char *path);
...
+ * @retval <0 negative errno code on error.
  */
 int fs_rename(const char *from, const char *to);
 
 /**
- * @brief File read
+ * @brief Read file
  *
- * Reads items of data of size bytes long.
+ * Reads @p size bytes of data to @p ptr pointed buffer, returns number of bytes
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:282]: Agree, thanks for the hint.

File: include/fs/fs.h [hunk 8]
```diff
@@ -260,115 +265,117 @@ int fs_unlink(const char *path);
...
 
 /**
- * @brief File write
+ * @brief Write file
  *
- * Writes items of data of size bytes long.
+ * Attempts to write @p size number of bytes to specified file.
+ * If negative value is returned from function, the file pointer has not
+ * been advanced.
+ * If function returns non-negative number that is less than @p size, the errno
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:301]: Will fix.

File: include/fs/fs.h [hunk 9]
```diff
@@ -260,115 +265,117 @@ int fs_unlink(const char *path);
...
+ * @retval 0 on success;
+ * @retval <0 negative errno code on error.
  */
 int fs_seek(struct fs_file_t *zfp, off_t offset, int whence);
 
 /**
  * @brief Get current file position.
  *
- * Retrieves the current position in the file.
+ * Retrieves and returns the current position in the file stream;  note that
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:336]: Ah, yes. Need to reword that. What I have tried to solve is following misunderstanding from one of devs: the person has been using fs_stat after appending data to file and did not understood why fs_tell and fs_stat give different values.

File: include/fs/fs.h [hunk 10]
```diff
@@ -260,115 +265,117 @@ int fs_unlink(const char *path);
...
 /**
- * @brief File write
+ * @brief Write file
  *
- * Writes items of data of size bytes long.
+ * Attempts to write @p size number of bytes to specified file.
+ * If negative value is returned from function, the file pointer has not
+ * been advanced.
+ * If function returns non-negative number that is less than @p size, the errno
+ * should be checked for error code, possibly the device has ran out of free
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:302]: OK, will fix.

File: include/fs/fs.h [hunk 11]
```diff
@@ -261,118 +265,120 @@ int fs_unlink(const char *path);
...
  */
 ssize_t fs_read(struct fs_file_t *zfp, void *ptr, size_t size);
 
 /**
- * @brief File write
+ * @brief Write file
  *
- * Writes items of data of size bytes long.
+ * Attempts to write @p size number of bytes to specified file.
+ * If negative value is returned from function, the file pointer has not
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h:299]: OK

File: include/fs/fs.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply include/fs/fs.h:336]: OK; maybe a note in `fs_stat` that the state may not be updated while the file is open, if there isn't something there.  (I'm not sure if sync is sufficient in all cases.)

File: include/fs/fs.h (hunk 5 above)
Comment (de-nordic) [reply include/fs/fs.h:501]: @de-nodic facepalms trying to figure out how he did missed that.

File: include/fs/fs.h [hunk 12]
```diff
@@ -261,118 +265,120 @@ int fs_unlink(const char *path);
...
  *
- * This function can be used to flush the cache of an open file. This can
- * be called to ensure data gets written to the storage media immediately.
- * This may be done to avoid data loss if power is removed unexpectedly.
- * Note that closing a file will cause caches to be flushed correctly so it
- * need not be called if the file is being closed.
+ * The function flushes the cache of an open file; it can be invoked to ensure
+ * data gets written to the storage media immediately, e.g. to avoid data loss
+ * in case if power is removed unexpectedly.
+ * @note It is not needed to call the function before @c fs_close.
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply include/fs/fs.h]: Brought back the original note.

**Comment by de-nordic** [comment]
@pabigot Shouldn't the parameter lists be also altered, example from `fs_read`, in the following way:
```
 * @param zfp Pointer to the file object
 * @param ptr Pointer to the data buffer
 * @param size Number of bytes to be read
```
to
```
 * @param zfp A pointer to the file object
 * @param ptr A pointer to the data buffer
 * @param size The number of bytes to be read
```
?

**Comment by pabigot** [comment]
> @pabigot Shouldn't the parameter lists be also altered, example from `fs_read`, in the following way:
> 
> ```
>  * @param zfp Pointer to the file object
>  * @param ptr Pointer to the data buffer
>  * @param size Number of bytes to be read
> ```
> 
> to
> 
> ```
>  * @param zfp A pointer to the file object
>  * @param ptr A pointer to the data buffer
>  * @param size The number of bytes to be read
> ```

You're right, there's a slippery slope here.  I seem to draw a distinction between regular prose, i.e. a significant amount of text consisting of full sentences formatted as a paragraph that is meant to be read as a unit, and a fragment documenting an identifier.  In the latter I would not capitalize the fragment, and would not use closing punctuation.

So were I writing the parameter section I'd do:

```
 * @param zfp pointer to the file object
 * @param ptr pointer to the data buffer
 * @param size number of bytes to be read
```

If you look at the [generated documentation](https://docs.zephyrproject.org/latest/reference/file_system/index.html#c.fs_read) you can see how the two contexts are different.

However again, I don't think there are rules, and at least some things do turn the fragments into full sentences, e.g. [this random selection](https://docs.zephyrproject.org/latest/reference/kernel/other/polling.html#c.k_poll_signal_raise).

So if you want to use sentences for the parameter descriptions, yes, "a" and "the" should be present and the sentence should be capitalized and have a period at the end.

I really wish we had a documentation maintainer who could provide direction on this sort of thing.

**Comment by de-nordic** [comment]
> 
> ```
>  * @param zfp pointer to the file object
>  * @param ptr pointer to the data buffer
>  * @param size number of bytes to be read
> ```
> 
This looks nice. I will not make changes to parameters list for now, it makes no sense as long as it is not specified somewhere.

Later on I will create RFC for gathering comments on API documentation formatting. Maybe after gathering some input we could decide on some common style.

> I really wish we had a documentation maintainer who could provide direction on this sort of thing.

Maybe some technical writer could help to at least set the style.

Btw. @pabigot thanks for committing your time to this boring review and fixing my English.

**Review by pabigot — APPROVED** [review]
Thanks, I think this is a definite improvement.

**Review by nvlsianpu — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**